server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -442,9 +444,20 @@ public ClusterState addIndexTemplateV2(final ClusterState currentState, final bo
 
         validate(name, finalIndexTemplate);
         logger.info("adding index template [{}]", name);
-        return ClusterState.builder(currentState)
+        ClusterState newState = ClusterState.builder(currentState)
             .metadata(Metadata.builder(currentState.metadata()).put(name, finalIndexTemplate))
             .build();
+        if (finalIndexTemplate.getDataStreamTemplate() != null) {
+            validateDataStreamTemplate(name, finalIndexTemplate, newState);
+        }
+        return newState;
+    }
+
+    private void validateDataStreamTemplate(String name, ComposableIndexTemplate finalIndexTemplate, ClusterState state) throws Exception {
+        String tsFieldName = finalIndexTemplate.getDataStreamTemplate().getTimestampField();
+        Map<String, Object> finalMapping = parseV2Mappings("{}", resolveMappings(state, name), xContentRegistry);
```

Comment:
When moving the validation to the `validateCompositeTemplate` method we should be able to reuse the mapping generated in that method